### PR TITLE
[Mobile Payments] Consolidate CardPresent and PaymentGateway Stores; Switch Between WCPay and Stripe Backends Within

### LINKED ISSUE #5531: [Mobile Payments] Update the app to use /account endpoint of the active plugin (WCPay vs Stripe Extension)
Update the app to use /account endpoint of the active plugin (WCPay vs Stripe Extension).

### LINKED ISSUE #5534: [Mobile Payments] Update the app to use /terminal/locations/store endpoint of the active plugin (WCPay vs Stripe Extension)
Update the app to use /terminal/locations/store endpoint of the active plugin (WCPay vs Stripe Extension).

Depends on https://github.com/woocommerce/woocommerce-ios/issues/5533

### LINKED ISSUE #5536: [Mobile Payments] Update the app to use /payments/connection_tokens endpoint of the active plugin (WCPay vs Stripe Extension)
Update the app to use /payments/connection_tokens endpoint of the active plugin (WCPay vs Stripe Extension).

Depends on #5535

### LINKED ISSUE #5853: [Mobile Payments] Only attempt to fetch payment gateway accounts for active extensions
https://github.com/woocommerce/woocommerce-ios/pull/5734 incorporated a change where PaymentGatewayAccountStore::loadAccounts attempts to load both WooCommerce Payments extension account details AND Stripe extension account details regardless of which extension is active. A better implementation would only fire the appropriate request.

### PR DESCRIPTION
**Pull request by allendav**
Closes: #5853
Closes: #5534
Closes: #5536
Closes: #5531

### Description
- Combines PaymentGatewayAccountStore and CardPresentPaymentStore, moving the actions in the former into the latter
- Consolidates their unit tests as well
- Changes onboarding to defer loading accounts until after it knows which plugin is active
- Also tells the CardPresentPaymentStore which backend to use for things like fetching accounts, connecting readers
- Defaults to WCPay backend

### Unfinished Business
- Define a protocol for the two remotes to simplify the code in CardPresentPaymentStore
- Add unit tests to cover the logic added to CardPresentPaymentStore and CardPresentPaymentsOnboardingUseCase - see also #5866 and #5867
- **I'm going to move these into subsequent PR(s) - this one is already so big.**

### Please note
- This does NOT hook up these changes to the Order flow. Capturing payments through Stripe powered stores will come in a later PR.

### Testing instructions
- First, ensure you can connect to a reader on a WCPay powered store
- Next, set up a store (if you haven't already) using the Stripe extension.
- IMPORTANT: Modify woocommerce-gateway-stripe/includes/admin/class-wc-rest-stripe-account-controller.php so that get_account_summary also returns keys of is_live (false) and test_mode (true) if necessary - see https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2262 for more details
- Enable the Stripe IPP feature under Settings > Experimental
- Then, ensure you can connect to a reader on that Stripe powered store

### Screenshots
N/A

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by ctarda** [comment]
It is looking pretty good. 

I figure you haven't reached this point yet, but just to make sure we don't miss it:
* when calling startDiscovery in the reader service, we pass a token provider that wraps the remote. That token provider will need to be different for Stripe and WCPay
* when calling reset into the service, we would need to do something, perhaps point the store to a default remote?

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingUseCase.swift
Comment (ctarda) [inline WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingUseCase.swift]: This could be removed

File: Yosemite/Yosemite/Actions/CardPresentPaymentAction.swift
Comment (ctarda) [inline Yosemite/Yosemite/Actions/CardPresentPaymentAction.swift:24]: We might want to revisit the naming of WCPayCustomer in the future?

File: Yosemite/Yosemite/Actions/CardPresentPaymentAction.swift
Comment (ctarda) [inline Yosemite/Yosemite/Actions/CardPresentPaymentAction.swift:9]: I don't mind declaring these two cases. It makes the code extremely declarative. 

One potential way to make this a bit more abstract could be to declare an enumeration (i.e. PaymentGateway) with two cases, and rename these two actions into a single action: `useGateway(gateway: PaymentGateway)

It could kind of future proof the API for a potential third gateway. On the other hand, a third gateway does not sound like something we would add in the nearly future, so 🤷🏼

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (ctarda) [inline Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift:128]: You mentioned in the description making both remotes implement a common interface in a future PR.

I would take that one step further, and I would suggest extracting the logic to decide which backend we are calling into to a separate entity. For example, that could be a class with a single public method similar to this (naming is terrible, but just to illustrate the idea):
```
func getRemote(gateway: PaymentGateway) ->  PaymentsRemote {
switch gateway {
case .stripe: return stripeRemote
case .wcpay: return wcpayRemote
}
}
```

This class could be declared within this store, so it is completely private.

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (ctarda) [inline Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift]: It would be good to remove these prints, if not now, in a follow-up PR

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (ctarda) [inline Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift]: I am not sure I would call reset here. If we do, the API docs of these methods should, in my opinion, clarify that there is going to be an implicit call to reset.

I think it would be slightly preferable to leave the responsibility of resetting the SDK to an upper layer.

Also, the Terminal SDK tends to error quite often for attempting operations when it's busy, so we should make sure we do not introduce instability

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (ctarda) [inline Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift]: Please don't make this a TODO, make it a ticket. Even if clients don't use it, it is good to surface as many errors as possible

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (ctarda) [inline Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift]: Same comment about making this a ticket instead of a TODO

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (ctarda) [inline Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift:130]: I would feel more comfortable merging this into trunk if we checked the feature flag in all these methods, and short-circuited defaulting to using the wcpayRemote when the feature flag is not active.

**Review by ctarda — APPROVED** [review]
Thanks for tackling this @allendav . I appreciate the iterative approach and how you have tried to find a reasonable solution given the trade-offs.

I left some comments for you to consider. The only thing I will suggest doing before merging is short-circuiting all the operations, defaulting to use WCPayRemote, when the feature flag is not enabled.

In that sense, extracting the logic to select the remote that should be used to a different class before merging, as I mentioned in a comment, might be helpful. But I'll leave that to you.

In any case, if you agree, I think it would be good to at least take the feature flag into consideration: safety first!

:shipit:

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (allendav) [reply Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift]: Oops. Meant to remove all of them.

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (allendav) [reply Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift]: I just went ahead and brought back failure - https://github.com/woocommerce/woocommerce-ios/pull/5868/commits/5c19233add147d6e8ac7b59c98533e66999fe17e

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (allendav) [reply Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift]: I just went ahead and brought back failure - https://github.com/woocommerce/woocommerce-ios/pull/5868/commits/5c19233add147d6e8ac7b59c98533e66999fe17e

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (allendav) [reply Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift]: Fixed in https://github.com/woocommerce/woocommerce-ios/pull/5868/commits/d8db38f438a71f5bc36e6b439dc882a2338ca6c1

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingUseCase.swift
Comment (allendav) [reply WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingUseCase.swift]: Fixed in https://github.com/woocommerce/woocommerce-ios/pull/5868/commits/f2f6a89eeaf4b89edace55186e922f83aacbd46a

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (allendav) [reply Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift:130]: Fixed in https://github.com/woocommerce/woocommerce-ios/pull/5868/commits/58187d63416200c480bdb8ab75e71be097adcbc1

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (allendav) [reply Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift]: I've removed the resets from the service layer in https://github.com/woocommerce/woocommerce-ios/pull/5868/commits/fc0132753e9c15b45801a177b871b6d904fdc9a1 but I haven't added them to the upper layer yet. During testing I noticed that even when switching between WCPay powered and Stripe powered stores, the Terminal SDK was continuing to ask the previous service for tokens and location.

I'll continue digging into this. I think we might have missed something in https://github.com/woocommerce/woocommerce-ios/pull/4147

File: Yosemite/Yosemite/Actions/CardPresentPaymentAction.swift
Comment (allendav) [reply Yosemite/Yosemite/Actions/CardPresentPaymentAction.swift:9]: Good idea, but let's cross that bridge if/when we get to a third gateway? :)

File: Yosemite/Yosemite/Actions/CardPresentPaymentAction.swift
Comment (allendav) [reply Yosemite/Yosemite/Actions/CardPresentPaymentAction.swift:24]: Opened https://github.com/woocommerce/woocommerce-ios/issues/5903

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (allendav) [reply Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift:128]: Opened https://github.com/woocommerce/woocommerce-ios/issues/5904

**Comment by allendav** [comment]
@ctarda - thank you for your review - I've addressed all the feedback but I am now going to see if this is a problem apart from this PR: https://github.com/woocommerce/woocommerce-ios/pull/5868#discussion_r785181633 (which I suspect it is)

If that turns out to be the case, we would be able to move forward with this PR and address that bug separately

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (ctarda) [reply Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift]: Could it because there is no call to reset() before switching the store?

Looking at the implementation of setConfigProvider in StripeCardReaderService, there is this:

```
        if !Terminal.hasTokenProvider() {
            Terminal.setTokenProvider(tokenProvider)
        }
```

**Comment by ctarda** [comment]
@allendav  can we please merge this as it is, regardless of [this ](https://github.com/woocommerce/woocommerce-ios/pull/5868#discussion_r785181633) and my reply to it? 

This PR is kind becoming a blocker for making further progress

File: Yosemite/Yosemite/Actions/CardPresentPaymentAction.swift
Comment (ctarda) [reply Yosemite/Yosemite/Actions/CardPresentPaymentAction.swift:9]: Right. There is one more argument supporting doing this soon, without waiting for a third plugin: serializing an enumeration to user defaults is simple. Have in mind we are going to have to save the plugin selected by the user to user defaults, and materialize it again from user defaults before performing any other operations. We could just save and read a boolean flag, but abstracting that behind an enumeration is very low effort for a significant benefit, imo.

So, sure, let's defer it, but not for too long.

**Comment by allendav** [comment]
Merged!

**Pull request merged**
